### PR TITLE
Ensure OutboundTcpStreamActor's promises are completed if the connection fails

### LINKED ISSUE #18151: OutboundTcpStreamActor does not fulfull promises if connection timesout
If a connection isn't established before the timeout, TcpStreamActor#handleSubscriptionTimeout will be called which will stop the actor without fulfilling the `processorPromise` or `localAddressPromise` promises. 

An unfulfilled `processorPromise` will cause the stream to stall. In the context of the superPool, the entire superPool will stall after this happens a few times.

### PR DESCRIPTION
**Pull request by ryandbair**
Fixes #18151

### CODE REVIEW

**Comment by typesafehub-validator** [comment]
Hi @ryandbair,

Thank you for your contribution! We really value the time you've taken to put this together.

We see that you have signed the Typesafe Contributors License Agreement before, however, the CLA has changed since you last signed it.
Please review the new CLA and sign it before we proceed with reviewing this pull request:

[http://www.typesafe.com/contribute/cla](http://www.typesafe.com/contribute/cla)

File: akka-samples/akka-docs-java-lambda/build.sbt [hunk 1]
```diff
@@ -15,7 +15,7 @@ val publishedAkkaVersion = "2.3.11"
 libraryDependencies ++= Seq(
   "com.typesafe.akka" %%      "akka-actor"                       % publishedAkkaVersion,
   "com.typesafe.akka" %%      "akka-testkit"                     % publishedAkkaVersion % "test",
-  "com.typesafe.akka" %%      "akka-stream-experimental"         % "1.0-SNAPSHOT",
-  "com.typesafe.akka" %%      "akka-stream-testkit-experimental" % "1.0-SNAPSHOT" % "test",
```
Comment (drewhk) [inline akka-samples/akka-docs-java-lambda/build.sbt]: please don't change this

File: akka-samples/akka-docs-java-lambda/pom.xml [hunk 2]
```diff
@@ -28,12 +28,12 @@
     <dependency>
       <groupId>com.typesafe.akka</groupId>
       <artifactId>akka-stream-experimental_2.10</artifactId>
-      <version>1.0-SNAPSHOT</version>
```
Comment (drewhk) [inline akka-samples/akka-docs-java-lambda/pom.xml]: please don't change this

File: project/AkkaBuild.scala [hunk 3]
```diff
@@ -40,7 +40,7 @@ object AkkaBuild extends Build {
 
   val requestedScalaVersion = System.getProperty("akka.scalaVersion", "2.10.5")
   val Seq(scalaEpoch, scalaMajor) = """(\d+)\.(\d+)\..*""".r.unapplySeq(requestedScalaVersion).get.map(_.toInt)
-  val streamAndHttpVersion = "1.0-SNAPSHOT"
```
Comment (drewhk) [inline project/AkkaBuild.scala]: please don't change this

Comment (ryandbair) [inline project/AkkaBuild.scala]: Those changes came in from the releasing-akka-stream-and-http-experimental-1.0 branch. Should I rebase on release-2.3-dev?

Comment (ktoso) [inline project/AkkaBuild.scala]: Yup - no need to touch anything from the other branches, just `release-2.3-dev`

**Comment by ktoso** [comment]
The fix LGTM; PR needs to not touch the version numbers and we can merge :)

**Comment by typesafehub-validator** [comment]
Hi @ryandbair,

Thank you for your contribution! We really value the time you've taken to put this together.

We see that you have signed the Typesafe Contributors License Agreement before, however, the CLA has changed since you last signed it.
Please review the new CLA and sign it before we proceed with reviewing this pull request:

[http://www.typesafe.com/contribute/cla](http://www.typesafe.com/contribute/cla)

**Comment by jrudolph** [comment]
How about adding some tests?

**Comment by drewhk** [comment]
+1 for adding tests. Otherwise LGTM

**Comment by ryandbair** [comment]
Tests are good. 

Any thoughts how to test this behavior? I rather like that the current tests don't go too deep into the internals, but I'm not sure how to test this otherwise. I've seen some tests that use an actor selection to look for actors which would work but that feels hacky.

**Comment by drewhk** [comment]
@ryandbair I would like to merge this, but you will need to sign the CLA. You can do it with your GitHub account so it is really just a few clicks. 

Thank you!

**Comment by ryandbair** [comment]
I had signed it after submitting the PR initially. I just re-signed it since it doesn't look like the bot recognized it the first time.

**Comment by drewhk** [comment]
Thanks, I checked manually now. The only change needed for merging this is to change to commit message to:

    =str #18151: Fail Futures returned by TCP when connectino fails

and then I merge after validation

**Comment by rkuhn** [comment]
OK TO TEST

**Comment by drewhk** [comment]
Pdf creation failed, unrelated.

**Pull request merged**